sqlmesh/core/schema_diff.py

```diff
@@ -14,83 +17,366 @@ class SchemaDeltaOp(Enum):
     DROP = auto()
     ALTER_TYPE = auto()
 
+    @property
+    def is_add(self) -> bool:
+        return self == SchemaDeltaOp.ADD
+
+    @property
+    def is_drop(self) -> bool:
+        return self == SchemaDeltaOp.DROP
+
+    @property
+    def is_alter_type(self) -> bool:
+        return self == SchemaDeltaOp.ALTER_TYPE
+
+
+class ParentColumns(PydanticModel):
+    columns: t.List[t.Tuple[str, exp.DataType]]
+
+    @classmethod
+    def create(cls, *columns_and_types: t.Tuple[str, t.Union[str, exp.DataType]]) -> ParentColumns:
+        columns_and_data_types = [
+            (column_name, exp.DataType.build(column_type))
+            for column_name, column_type in columns_and_types
+        ]
+        return cls(columns=columns_and_data_types)
+
+    @classmethod
+    def empty(cls) -> ParentColumns:
+        return cls(columns=[])
+
+    def add(self, name: str, type: exp.DataType) -> ParentColumns:
+        # We don't need the full information for rested types, just the fact that they are nested
+        if type.this == exp.DataType.Type.STRUCT:
+            type = exp.DataType.build("STRUCT")
+        if type.this == exp.DataType.Type.ARRAY:
+            type = exp.DataType.build("ARRAY")
+        return ParentColumns(columns=self.columns + [(name, type)])
+
+    @property
+    def has_columns(self) -> bool:
+        return len(self.columns) > 0
+
+    def sql(self, array_suffix: t.Optional[str] = None) -> str:
+        return ".".join(
+            [
+                c[0]
+                if c[1].this != exp.DataType.Type.ARRAY
+                else f"{c[0]}{array_suffix if array_suffix else ''}"
+                for c in self.columns
+            ]
+        )
+
+
+class ColumnPosition(PydanticModel):
+    is_first: bool
+    is_last: bool
+    after: t.Optional[str] = None
+
+    @classmethod
+    def create_first(cls) -> ColumnPosition:
+        return cls(is_first=True, is_last=False, after=None)
+
+    @classmethod
+    def create_last(cls, after: t.Optional[str] = None) -> ColumnPosition:
+        return cls(is_first=False, is_last=True, after=after)
+
+    @classmethod
+    def create_middle(cls, after: str) -> ColumnPosition:
+        return cls(is_first=False, is_last=False, after=after)
+
+    @classmethod
+    def create(
+        cls,
+        pos: int,
+        current_kwargs: t.List[exp.StructKwarg],
+        replacing_col: bool = False,
+    ) -> ColumnPosition:
+        is_first = pos == 0
+        is_last = pos == len(current_kwargs) - int(replacing_col)
+        after = None
+        if not is_first:
+            prior_kwarg = current_kwargs[pos - 1]
+            after, _ = _get_name_and_type(prior_kwarg)
+        return cls(is_first=is_first, is_last=is_last, after=after)
+
+    @property
+    def sqlglot_col_position(self) -> t.Optional[exp.ColumnPosition]:
```

Comment:
why is this called sqlglot_col_position?

Reply:
id' prefer not to prefix it with sqlglot, seems unnecessary especially given the type hints

Reply:
SQLMesh also has a `ColumnPosition` class so making it clear we are returning the SQLGlot expression of `ColumnPosition`.

Reply:
maybe col it col_position_node or ast or something